// tmpl-log/lib/const.js
exports.DEFAULT_TAGS = {
  "default": [],
  bold: ['\x1B[1m'],
  italic: ['\x1B[3m'],
  underline: ['\x1B[4m'],
  inverse: ['\x1B[7m'],
  strikethrough: ['\x1B[9m'],
  white: ['\x1B[37m'],
  grey: ['\x1B[90m'],
  black: ['\x1B[30m'],
  blue: ['\x1B[34m'],
  cyan: ['\x1B[36m'],
  green: ['\x1B[32m'],
  magenta: ['\x1B[35m'],
  red: ['\x1B[31m'],
  yellow: ['\x1B[33m'],
  whiteBG: ['\x1B[47m'],
  greyBG: ['\x1B[49;5;8m'],
  blackBG: ['\x1B[40m'],
  blueBG: ['\x1B[44m'],
  cyanBG: ['\x1B[46m'],
  greenBG: ['\x1B[42m'],
  magentaBG: ['\x1B[45m'],
  redBG: ['\x1B[41m'],
  yellowBG: ['\x1B[43m']
};

exports.CLOSE_CODES = {
  '\x1B[1m': '\x1B[22m',
  '\x1B[3m': '\x1B[23m',
  '\x1B[4m': '\x1B[24m',
  '\x1B[7m': '\x1B[27m',
  '\x1B[9m': '\x1B[29m',
  '\x1B[37m': '\x1B[39m',
  '\x1B[90m': '\x1B[39m',
  '\x1B[30m': '\x1B[39m',
  '\x1B[34m': '\x1B[39m',
  '\x1B[36m': '\x1B[39m',
  '\x1B[32m': '\x1B[39m',
  '\x1B[35m': '\x1B[39m',
  '\x1B[31m': '\x1B[39m',
  '\x1B[33m': '\x1B[39m',
  '\x1B[47m': '\x1B[49m',
  '\x1B[49;5;8m': '\x1B[49m',
  '\x1B[40m': '\x1B[49m',
  '\x1B[44m': '\x1B[49m',
  '\x1B[46m': '\x1B[49m',
  '\x1B[42m': '\x1B[49m',
  '\x1B[45m': '\x1B[49m',
  '\x1B[41m': '\x1B[49m',
  '\x1B[43m': '\x1B[49m'
};
